anta/result_manager/models.py

```diff
@@ -111,16 +104,103 @@ def _set_status(self, status: AntaTestStatus, message: str | None = None) -> Non
             Status of the test.
         message
             Optional message.
+        """
+        self.result = status
+        if message is not None:
+            self.messages.append(message)
+
 
+class AtomicTestResult(BaseTestResult):
+    """Describe the result of an atomic test part of a larger test related to a TestResult instance.
+
+    Attributes
+    ----------
+    description : str | None
+        Description of the AtomicTestResult.
+    inputs: BaseModel | None
+        If this AtomicTestResult is related to a specific parent test input, this field must be set.
+    result : AntaTestStatus
+        Result of the atomic test.
+    messages : list[str]
+        Messages reported by the test.
+    """
+
+    _parent: TestResult
+
+    def __init__(self, **data: Any) -> None:  # noqa: ANN401
+        """Instantiate the parent TestResult private attribute."""
+        parent = data.pop("parent")
+        super().__init__(**data)
+        self._parent = parent
+
+    def _set_status(self, status: AntaTestStatus, message: str | None = None) -> None:
+        """Set status and insert optional message.
+
+        Parameters
+        ----------
+        status
+            Status of the test.
+        message
+            Optional message.
         """
         self.result = status
+        if (self._parent.result == AntaTestStatus.UNSET and status == AntaTestStatus.SUCCESS) or status in [AntaTestStatus.FAILURE, AntaTestStatus.ERROR]:
+            self._parent.result = status
         if message is not None:
             self.messages.append(message)
+            self._parent.messages.append(message)
+
+
+class TestResult(BaseTestResult):
+    """Describe the result of a test from a single device.
+
+    Attributes
+    ----------
+    name : str
+        Name of the device on which the test was run.
+    test : str
+        Name of the AntaTest subclass.
+    inputs:  BaseModel | None
+        Inputs of the AntaTest instance.
+    categories : list[str]
+        List of categories the TestResult belongs to. Defaults to the AntaTest subclass categories.
+    description : str
+        Description of the TestResult. Defaults to the AntaTest subclass description.
+    result : AntaTestStatus
+        Result of the test.
+    messages : list[str]
+        Messages reported by the test.
+    custom_field : str | None
+        Custom field to store a string for flexibility in integrating with ANTA.
+    atomic_results: list[AtomicTestResult]
+        A list of AtomicTestResult instances which can be used to store atomic results during the test execution.
+        It can then be leveraged in the report to render atomic results over the test global TestResult.
```

Comment:
When serializing a TestResult, the order of the fields seem to be following the class hierchary, `description`, `result` are at the top which is probably not what we want.

Reply:
There is no way to do that with field inheritance.
So I guess we will redefine the fields in both models.